contracts/src/v0.1/L1EndpointRequestResponse.sol

```diff
@@ -0,0 +1,301 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+import "./RequestResponseConsumerBase.sol";
+import "./L1EndpointBase.sol";
+import "./RequestResponseConsumerFulfill.sol";
+
+abstract contract L1EndpointRequestResponse is
+    RequestResponseConsumerBase,
+    L1EndpointBase,
+    RequestResponseConsumerFulfillUint128,
+    RequestResponseConsumerFulfillInt256,
+    RequestResponseConsumerFulfillBool,
+    RequestResponseConsumerFulfillString,
+    RequestResponseConsumerFulfillBytes32,
+    RequestResponseConsumerFulfillBytes
+{
+    using Orakl for Orakl.Request;
+
+    event DataRequested(uint256 requestId, address sender);
+    event DataRequestFulfilled(
+        uint256 requestId,
+        uint256 l2RequestId,
+        address sender,
+        uint256 callbackGasLimit,
+        bytes32 jobId,
+        uint128 responseUint128,
+        int256 responseInt256,
+        bool responseBool,
+        string responseString,
+        bytes32 responseBytes32,
+        bytes responseBytes
+    );
+
+    constructor(
+        address requestResponseCoordinator
+    ) RequestResponseConsumerBase(requestResponseCoordinator) {}
+
+    function requestData(
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint8 numSubmission,
+        address sender,
+        uint256 l2RequestId,
+        Orakl.Request memory req
+    ) internal returns (uint256) {
+        if (!sOracles[msg.sender]) {
+            revert OnlyOracle();
+        }
+        //check consumer and balance
+        bool isValidConsumer = REGISTRY.isValidConsumer(accId, sender);
+        if (!isValidConsumer) {
+            revert ConsumerValid();
+        }
+        uint256 balance = REGISTRY.getBalance(accId);
+        uint64 reqCount = 0;
+        uint256 fee = COORDINATOR.estimateFee(reqCount, 1, callbackGasLimit);
+        REGISTRY.decreaseBalance(accId, fee);
+        if (balance < fee) {
+            revert InsufficientBalance();
+        }
+        uint256 id = COORDINATOR.requestData{value: fee}(
+            req,
+            callbackGasLimit,
+            numSubmission,
+            address(this)
+        );
+        sRequest[id] = RequestDetail(l2RequestId, sender, callbackGasLimit);
+        emit DataRequested(id, sender);
+        return id;
+    }
+
+    //request data
+    function requestDataUint128(
```

Comment:
I don't think we need the following functions, because all requests can be called through `requestData` function.

* `requestDataUint128`
* `requestDataInt256`
* `requestDataBool`
* `requestDataString`
* `requestDataBytes32`
* `requestDataBytes`

Reply:
The reason why we need this is we need build requestID onchain using the `Orakl.Request` library. For each type of data the request object is built differently:
`Orakl.Request memory req = buildRequest(jobId);`
If we consolidate the functions we need to build the request object offchain, which will make the whole thing more complicated.

Reply:
At what point do consumers specify their request?

Reply:
This called by our reporter on L1.
Listen on L2 --> reporter call this function on L1.

Reply:
@KelvinThai I think we misunderstand each other. The `requestDataUint128` and all other functions defined for other data types include request describing to fetch price from https://min-api.cryptocompare.com which looks like we ignore any kind of request from L2. Is that on purpose?

Reply:
aha! so you mean we need to find other APIs that actually returns other datatypes? That one API is the dummy one we used in all testcases and demos

Reply:
The request has to be specified by user through smart contract on L2. There should be no connection to any predefined 3rd party API. Request-Response is general solution that allows to fetch any type of data from any 3rd party API that the user wants to use.

Reply:
I got it now, after getting back to the code. I'll fix it.